source/Halibut.Tests/BiDirectionalTests.cs

```diff
@@ -0,0 +1,110 @@
+    {
+        /// <summary>
+        /// Test is redundant but kept around since we really want the security part to work. 
+        /// </summary>
+        [Test]
+        public async Task ClientCanRespondToServerRequests()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var octopusResponse = await octopusEchoClient.SayHelloAsync("Hello");
+                var tentacleResponse = await tentacleEchoClient.SayHelloAsync("World");
+
+                octopusResponse.Should().Be("Hello...");
+                tentacleResponse.Should().Be("World...");
+            });
+        }
+
+        [Test]
+        public async Task ConcurrentTasksDoNotBlock()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var sw = Stopwatch.StartNew();
+                var t1 = Task.Run(async () => await RunLongRunningTask(octopusEchoClient)).WithTimeout(TimeSpan.FromSeconds(12));
+                var t2 = Task.Run(async () => await RunLongRunningTask(tentacleEchoClient)).WithTimeout(TimeSpan.FromSeconds(12));
+
+                await Task.WhenAll(t1, t2);
+                sw.Stop();
+
+                sw.ElapsedMilliseconds.Should().BeLessThan(await t1 + await t2);
+            });
+            
+            async Task<long> RunLongRunningTask(IAsyncClientEchoService tentacleEchoClient)
+            {
+                var longRunningDuration = Stopwatch.StartNew();
+                await tentacleEchoClient.LongRunningOperationAsync();
+                longRunningDuration.Stop();
+                return longRunningDuration.ElapsedMilliseconds;
+            }
+        }
+
+        [Test]
+        public async Task ConcurrentClientAndServerRequestsCorrectlyInterleaved()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var t1 = Task.Run(async () => await RunEchoTask(octopusEchoClient)).WithTimeout(TimeSpan.FromSeconds(5));
+                var t2 = Task.Run(async () => await RunEchoTask(tentacleEchoClient)).WithTimeout(TimeSpan.FromSeconds(5));
+                
+                await Task.WhenAll(t1, t2);
+
+                t1.Status.Should().Be(TaskStatus.RanToCompletion);
+                t2.Status.Should().Be(TaskStatus.RanToCompletion);
+            });
+            
+            async Task RunEchoTask(IAsyncClientEchoService tentacleEchoClient)
+            {
+                 var j = 0;
+                 while (j++ < 10)
+                 {
+                     var tentacleResponse = await tentacleEchoClient.SayHelloAsync($"Hello{j}");
+                     tentacleResponse.Should().Be($"Hello{j}...");
+                 }
+            }
+        }
+        
+        async Task SetupBiDirectionalClients(Func<IAsyncClientEchoService, IAsyncClientEchoService, Task> thing)
```

Comment:
Nit: it would be great if this was made to be a builder like the builders used in other tests. However I don't suspect we will have lots more of these types of tests.

Reply:
Unless we start using this format elsewhere I'd probably err towards YAGNI otherwise its hard to gauge what the reusable parts are.